File: apps/web/src/lib/payment-links/service.ts
import { randomBytes } from "node:crypto";
import { and, asc, count, desc, eq, ilike, inArray, or, type SQL } from "drizzle-orm";
import { db } from "@/lib/db";
import {
  organizations,
  paymentLinkItems,
  paymentLinks,
  type Organization,
  type PaymentLink,
} from "@/lib/db/schema";
import { DEFAULT_AUTH_URL } from "@/constants/app";
import type { AllowedAsset } from "@/lib/assets/types";
import { resolveAssetConfig } from "@/lib/assets/config";
import {
  dbAllowedAssets,
  serializePaymentLinkAssets,
} from "@/lib/assets/serialize";
import {
  calculateInvoiceTotal,
  lineItemAmount,
  parseFiatAmount,
  type InvoiceLineItemInput,
} from "@/lib/invoices/amount";
import {
  isInvoiceCurrencyCode,
  type InvoiceCurrencyCode,
} from "@/lib/invoices/currencies";
import { PLACEHOLDER_PRICING_PAYMENT_AMOUNT } from "@/constants/payments/defaults";
import { organizationEnvironmentWhere } from "@/lib/organizations/environment-scope";
import { createPayment, getCheckoutUrl } from "@/lib/payments/service";
import { paymentLinkCustomerInputSchema } from "@/lib/payment-links/schemas";
import {
  DEFAULT_PAYMENT_LINK_CUSTOMER_COLLECTION,
  hasCustomerCollection,
  normalizePaymentLinkCustomerCollection,
  type PaymentLinkCustomerCollection,
  type PaymentLinkCustomerInput,
  type PaymentLinkLineItem,
} from "@/lib/payment-links/types";

function createLinkPublicId() {
  return `plink_${randomBytes(12).toString("base64url")}`;
}

export function getPaymentLinkUrl(publicId: string) {
  const baseUrl = process.env.AUTH_URL ?? DEFAULT_AUTH_URL;
  return `${baseUrl}/l/${publicId}`;
}

function serializeLineItems(
  items: Array<{
    description: string;
    quantity: string;
    unitAmount: string;
  }>,
  currencyCode: InvoiceCurrencyCode
): PaymentLinkLineItem[] {
  return items.map((item) => ({
    description: item.description,
    quantity: item.quantity,
    unit_amount: item.unitAmount,
    line_amount: lineItemAmount(
      {
        description: item.description,
        quantity: item.quantity,
        unitAmount: item.unitAmount,
      },
      currencyCode
    ),
  }));
}

export async function listPaymentLinkItems(paymentLinkId: string) {
  return db
    .select()
    .from(paymentLinkItems)
    .where(eq(paymentLinkItems.paymentLinkId, paymentLinkId))
    .orderBy(asc(paymentLinkItems.sortOrder), asc(paymentLinkItems.createdAt));
}

async function insertPaymentLinkItems(
  paymentLinkId: string,
  items: InvoiceLineItemInput[],
  currencyCode: InvoiceCurrencyCode
) {
  if (items.length === 0) {
    return [];
  }

  return db
    .insert(paymentLinkItems)
    .values(
      items.map((item, index) => ({
        paymentLinkId,
        description: item.description.trim(),
        quantity: item.quantity.trim(),
        unitAmount: parseFiatAmount(item.unitAmount.trim(), currencyCode),
        sortOrder: index,
      }))
    )
    .returning();
}

function buildCustomerMetadata(
  input: PaymentLinkCustomerInput | undefined,
  collection: PaymentLinkCustomerCollection
) {
  if (!input || !hasCustomerCollection(collection)) {
    return undefined;
  }

  const metadata: Record<string, string> = {};

  if (collection.collect_customer_name && input.customer_name?.trim()) {
    metadata.customer_name = input.customer_name.trim();
  }

  if (collection.collect_business_name && input.business_name?.trim()) {
    metadata.business_name = input.business_name.trim();
  }

  if (collection.require_phone_number && input.phone_number?.trim()) {
    metadata.phone_number = input.phone_number.trim();
  }

  if (collection.collect_customer_address) {
    const addressFields = [
      ["address_line1", input.address_line1],
      ["address_line2", input.address_line2],
      ["address_city", input.address_city],
      ["address_state", input.address_state],
      ["address_postal_code", input.address_postal_code],
      ["address_country", input.address_country],
    ] as const;

    for (const [key, value] of addressFields) {
      if (value?.trim()) {
        metadata[key] = value.trim();
      }
    }
  }

  return Object.keys(metadata).length > 0 ? metadata : undefined;
}

function validateCustomerInput(
  input: PaymentLinkCustomerInput | undefined,
  collection: PaymentLinkCustomerCollection
) {
  if (!hasCustomerCollection(collection)) {
    return;
  }

  const parsed = paymentLinkCustomerInputSchema.safeParse(input ?? {});

  if (!parsed.success) {
    throw new Error(parsed.error.issues[0]?.message ?? "Invalid customer details");
  }

  const data = parsed.data;

  if (collection.collect_customer_name && !data.customer_name?.trim()) {
    throw new Error("Customer name is required");
  }

  if (collection.collect_business_name && !data.business_name?.trim()) {
    throw new Error("Business name is required");
  }

  if (collection.require_phone_number && !data.phone_number?.trim()) {
    throw new Error("Phone number is required");
  }

  if (collection.collect_customer_address) {
    if (!data.address_line1?.trim()) {
      throw new Error("Address line 1 is required");
    }

    if (!data.address_city?.trim()) {
      throw new Error("City is required");
    }

    if (!data.address_country?.trim()) {
      throw new Error("Country is required");
    }
  }
}

export async function listPaymentLinks(
  organizationId: string,
  environment: Organization["environment"],
  limit = 50
) {
  return db
    .select()
    .from(paymentLinks)
    .where(
      organizationEnvironmentWhere(
        paymentLinks.organizationId,
        paymentLinks.environment,
        organizationId,
        environment
      )
    )
    .orderBy(desc(paymentLinks.createdAt))
    .limit(limit);
}

export type PaymentLinkSortOrder = "asc" | "desc";

export type ListPaymentLinksQuery = {
  page?: number;
  pageSize?: number;
  search?: string;
  status?: "active" | "inactive";
  sortOrder?: PaymentLinkSortOrder;
};

export async function listPaymentLinksPaginated(
  organizationId: string,
  environment: Organization["environment"],
  query: ListPaymentLinksQuery = {},
) {
  const page = Math.max(1, query.page ?? 1);
  const pageSize = Math.min(100, Math.max(1, query.pageSize ?? 20));
  const offset = (page - 1) * pageSize;
  const search = query.search?.trim();
  const sortOrder = query.sortOrder ?? "desc";

  const envWhere = organizationEnvironmentWhere(
    paymentLinks.organizationId,
    paymentLinks.environment,
    organizationId,
    environment,
  );

  let whereClause: SQL | undefined = envWhere;

  if (search) {
    const pattern = `%${search}%`;
    whereClause = and(
      whereClause,
      or(
        ilike(paymentLinks.publicId, pattern),
        ilike(paymentLinks.productName, pattern),
        ilike(paymentLinks.description, pattern),
      ),
    );
  }

  if (query.status === "active") {
    whereClause = and(whereClause, eq(paymentLinks.active, 1));
  } else if (query.status === "inactive") {
    whereClause = and(whereClause, eq(paymentLinks.active, 0));
  }

  const orderBy =
    sortOrder === "asc"
      ? asc(paymentLinks.createdAt)
      : desc(paymentLinks.createdAt);

  const [rows, totalRows] = await Promise.all([
    db
      .select()
      .from(paymentLinks)
      .where(whereClause)
      .orderBy(orderBy)
      .limit(pageSize)
      .offset(offset),
    db
      .select({ count: count() })
      .from(paymentLinks)
      .where(whereClause),
  ]);

  const linkIds = rows.map((row) => row.id);
  const itemCounts =
    linkIds.length > 0
      ? await db
          .select({
            paymentLinkId: paymentLinkItems.paymentLinkId,
            count: count(),
          })
          .from(paymentLinkItems)
          .where(inArray(paymentLinkItems.paymentLinkId, linkIds))
          .groupBy(paymentLinkItems.paymentLinkId)
      : [];

  const itemCountMap = new Map(
    itemCounts.map((row) => [row.paymentLinkId, row.count] as const),
  );

  const payment_links = await Promise.all(
    rows.map(async (link) => ({
      ...(await serializePaymentLink(link)),
      item_count: itemCountMap.get(link.id) ?? 0,
    })),
  );

  return {
    payment_links,
    total: totalRows[0]?.count ?? 0,
  };
}

export async function getPaymentLinkByPublicId(publicId: string) {
  const [link] = await db
    .select()
    .from(paymentLinks)
    .where(eq(paymentLinks.publicId, publicId))
    .limit(1);

  return link ?? null;
}

export async function getPaymentLinkForOrganization(
  publicId: string,
  organizationId: string,
  environment: Organization["environment"]
) {
  const link = await getPaymentLinkByPublicId(publicId);

  if (
    !link ||
    link.organizationId !== organizationId ||
    link.environment !== environment
  ) {
    return null;
  }

  return link;
}

export async function getPublicPaymentLinkDetail(publicId: string) {
  const link = await getPaymentLinkByPublicId(publicId);

  if (!link || !link.active) {
    return null;
  }

  const [organization] = await db
    .select({
      name: organizations.name,
      logoUrl: organizations.logoUrl,
      logoInitials: organizations.logoInitials,
    })
    .from(organizations)
    .where(eq(organizations.id, link.organizationId))
    .limit(1);

  if (!organization) {
    return null;
  }

  const customerCollection = normalizePaymentLinkCustomerCollection(
    link.customerCollection
  );

  const dbItems = await listPaymentLinkItems(link.id);
  const currencyCode = link.currencyCode ?? null;

  const items =
    currencyCode && isInvoiceCurrencyCode(currencyCode)
      ? serializeLineItems(
          dbItems.map((item) => ({
            description: item.description,
            quantity: item.quantity,
            unitAmount: item.unitAmount,
          })),
          currencyCode
        )
      : dbItems.map((item) => ({
          description: item.description,
          quantity: item.quantity,
          unit_amount: item.unitAmount,
          line_amount: item.unitAmount,
        }));

  return {
    id: link.publicId,
    amount: link.amount,
    currency_code: currencyCode,
    settlement_asset: {
      asset_code: link.settlementAsset,
      issuer_address: link.settlementAssetIssuer,
    },
    allowed_assets: link.allowedAssets ?? [],
    items,
    product_name: link.productName,
    product_description: link.productDescription,
    description: link.description,
    customer_collection: customerCollection,
    requires_customer_details: hasCustomerCollection(customerCollection),
    environment: link.environment,
    organization,
  };
}

export async function createPaymentLink(input: {
  organizationId: string;
  environment: Organization["environment"];
  currencyCode: InvoiceCurrencyCode;
  items: InvoiceLineItemInput[];
  settlementAsset?: AllowedAsset | null;
  allowedAssets?: AllowedAsset[] | null;
  description?: string | null;
  customerCollection?: PaymentLinkCustomerCollection;
  metadata?: Record<string, string> | null;
}) {
  if (!isInvoiceCurrencyCode(input.currencyCode)) {
    throw new Error("Unsupported payment link currency");
  }

  const amount = calculateInvoiceTotal(input.items, input.currencyCode);
  const assetConfig = await resolveAssetConfig({
    organizationId: input.organizationId,
    settlementAsset: input.settlementAsset,
    allowedAssets: input.allowedAssets,
  });

  const publicId = createLinkPublicId();
  const firstItem = input.items[0];

  const [link] = await db
    .insert(paymentLinks)
    .values({
      publicId,
      organizationId: input.organizationId,
      environment: input.environment,
      amount,
      currencyCode: input.currencyCode,
      settlementAsset: assetConfig.settlement_asset.asset_code,
      settlementAssetIssuer: assetConfig.settlement_asset.issuer_address,
      allowedAssets: dbAllowedAssets(assetConfig.allowed_assets),
      productName: firstItem?.description.trim() || null,
      productDescription: null,
      description: input.description?.trim() || null,
      customerCollection: normalizePaymentLinkCustomerCollection(
        input.customerCollection
      ),
      metadata: input.metadata ?? null,
      active: 1,
    })
    .returning();

  await insertPaymentLinkItems(link.id, input.items, input.currencyCode);

  return link;
}

export async function startCheckoutFromPaymentLink(
  publicId: string,
  customerInput?: PaymentLinkCustomerInput
) {
  const link = await getPaymentLinkByPublicId(publicId);

  if (!link || !link.active) {
    return null;
  }

  const customerCollection = normalizePaymentLinkCustomerCollection(
    link.customerCollection
  );

  validateCustomerInput(customerInput, customerCollection);

  const customerMetadata = buildCustomerMetadata(customerInput, customerCollection);
  const metadata = {
    ...(link.metadata ?? {}),
    ...(customerMetadata ?? {}),
    payment_link_id: link.publicId,
  };

  const usesFiatPricing = Boolean(link.currencyCode);

  const payment = await createPayment({
    organizationId: link.organizationId,
    environment: link.environment,
    amount: usesFiatPricing ? PLACEHOLDER_PRICING_PAYMENT_AMOUNT : link.amount,
    pricingCurrency: usesFiatPricing ? link.currencyCode : null,
    pricingAmount: usesFiatPricing ? link.amount : null,
    settlementAsset: {
      asset_code: link.settlementAsset,
      issuer_address: link.settlementAssetIssuer,
    },
    allowedAssets: link.allowedAssets ?? [],
    description: link.productName ?? link.description,
    metadata,
    sourceType: "payment_link",
    paymentLinkId: link.id,
  });

  await db
    .update(paymentLinks)
    .set({ updatedAt: new Date() })
    .where(eq(paymentLinks.id, link.id));

  return { link, payment, checkoutUrl: getCheckoutUrl(payment.publicId) };
}

export async function serializePaymentLink(
  link: PaymentLink,
  options?: { includeItems?: boolean }
) {
  const currencyCode = link.currencyCode ?? null;
  let items: PaymentLinkLineItem[] | undefined;

  if (options?.includeItems) {
    const dbItems = await listPaymentLinkItems(link.id);

    items =
      currencyCode && isInvoiceCurrencyCode(currencyCode)
        ? serializeLineItems(
            dbItems.map((item) => ({
              description: item.description,
              quantity: item.quantity,
              unitAmount: item.unitAmount,
            })),
            currencyCode
          )
        : dbItems.map((item) => ({
            description: item.description,
            quantity: item.quantity,
            unit_amount: item.unitAmount,
            line_amount: item.unitAmount,
          }));
  }

  return {
    id: link.publicId,
    object: "payment_link",
    amount: link.amount,
    currency_code: currencyCode,
    ...serializePaymentLinkAssets(link),
    product_name: link.productName,
    product_description: link.productDescription,
    items,
    description: link.description,
    customer_collection: normalizePaymentLinkCustomerCollection(
      link.customerCollection
    ),
    active: Boolean(link.active),
    metadata: link.metadata,
    url: getPaymentLinkUrl(link.publicId),
    environment: link.environment,
    created_at: link.createdAt,
    updated_at: link.updatedAt,
  };
}

export {
  DEFAULT_PAYMENT_LINK_CUSTOMER_COLLECTION,
  hasCustomerCollection,
  normalizePaymentLinkCustomerCollection,
};
export type {
  PaymentLinkCustomerCollection,
  PaymentLinkCustomerInput,
  PaymentLinkLineItem,
  PaymentLinkPresentation,
} from "@/lib/payment-links/types";

File: apps/api/internal/service/paymentlinks/service.go
package paymentlinks

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	invoicesvc "github.com/payoesteam/payoes/apps/api/internal/service/invoices"
	paymentmethodssvc "github.com/payoesteam/payoes/apps/api/internal/service/paymentmethods"
)

// ported from: apps/web/src/lib/payment-links/service.ts

type CustomerCollection struct {
	CollectCustomerName    bool `json:"collect_customer_name"`
	CollectBusinessName    bool `json:"collect_business_name"`
	CollectCustomerAddress bool `json:"collect_customer_address"`
	RequirePhoneNumber     bool `json:"require_phone_number"`
}

type PaymentLink struct {
	ID                    string
	PublicID              string
	OrganizationID        string
	Environment           string
	Amount                string
	CurrencyCode          *string
	SettlementAsset       string
	SettlementAssetIssuer *string
	AllowedAssets         []paymentmethodssvc.AllowedAsset
	Description           *string
	ProductName           *string
	ProductDescription    *string
	CustomerCollection    CustomerCollection
	Active                bool
	Metadata              map[string]string
	CreatedAt             time.Time
	UpdatedAt             time.Time
}

type Item struct {
	ID            string
	PaymentLinkID string
	Description   string
	Quantity      string
	UnitAmount    string
	SortOrder     int
	CreatedAt     time.Time
}

type ListQuery struct {
	Page      int
	PageSize  int
	Search    string
	Status    string
	SortOrder string
}

type CreateInput struct {
	OrganizationID     string
	Environment        string
	CurrencyCode       string
	Items              []invoicesvc.LineItemInput
	SettlementAsset    *paymentmethodssvc.AllowedAsset
	AllowedAssets      []paymentmethodssvc.AllowedAsset
	Description        *string
	CustomerCollection *CustomerCollection
	Metadata           map[string]string
}

type Service struct {
	pool           *pgxpool.Pool
	paymentMethods *paymentmethodssvc.Service
	webURL         string
}

func NewService(pool *pgxpool.Pool, paymentMethods *paymentmethodssvc.Service, webURL string) *Service {
	return &Service{pool: pool, paymentMethods: paymentMethods, webURL: webURL}
}

func createPublicID() (string, error) {
	buf := make([]byte, 12)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return "plink_" + base64.RawURLEncoding.EncodeToString(buf), nil
}

func (s *Service) URL(publicID string) string {
	return strings.TrimRight(s.webURL, "/") + "/c/" + publicID
}

func NormalizeCustomerCollection(v *CustomerCollection) CustomerCollection {
	if v == nil {
		return CustomerCollection{}
	}
	return CustomerCollection{
		CollectCustomerName:    v.CollectCustomerName,
		CollectBusinessName:    v.CollectBusinessName,
		CollectCustomerAddress: v.CollectCustomerAddress,
		RequirePhoneNumber:     v.RequirePhoneNumber,
	}
}

func (s *Service) scanLink(row pgx.Row) (*PaymentLink, error) {
	var (
		link          PaymentLink
		allowedRaw    []byte
		collectionRaw []byte
		metaRaw       []byte
		active        int
	)
	err := row.Scan(
		&link.ID, &link.PublicID, &link.OrganizationID, &link.Environment, &link.Amount,
		&link.CurrencyCode, &link.SettlementAsset, &link.SettlementAssetIssuer, &allowedRaw,
		&link.Description, &link.ProductName, &link.ProductDescription, &collectionRaw,
		&active, &metaRaw, &link.CreatedAt, &link.UpdatedAt,
	)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	link.Active = active == 1
	if len(allowedRaw) > 0 {
		_ = json.Unmarshal(allowedRaw, &link.AllowedAssets)
	}
	if len(collectionRaw) > 0 {
		_ = json.Unmarshal(collectionRaw, &link.CustomerCollection)
	}
	if len(metaRaw) > 0 {
		_ = json.Unmarshal(metaRaw, &link.Metadata)
	}
	return &link, nil
}

const linkSelect = `
	id, public_id, organization_id, environment, amount, currency_code,
	settlement_asset, settlement_asset_issuer, allowed_assets, description,
	product_name, product_description, customer_collection, active, metadata,
	created_at, updated_at`

func (s *Service) GetByPublicID(ctx context.Context, publicID string) (*PaymentLink, error) {
	return s.scanLink(s.pool.QueryRow(ctx, `
		SELECT `+linkSelect+` FROM payment_links WHERE public_id = $1 LIMIT 1`, publicID))
}

func (s *Service) GetForOrganization(ctx context.Context, publicID, organizationID, environment string) (*PaymentLink, error) {
	link, err := s.GetByPublicID(ctx, publicID)
	if err != nil || link == nil {
		return nil, err
	}
	if link.OrganizationID != organizationID || link.Environment != environment {
		return nil, nil
	}
	return link, nil
}

func (s *Service) ListItems(ctx context.Context, paymentLinkID string) ([]Item, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT id, payment_link_id, description, quantity, unit_amount, sort_order, created_at
		FROM payment_link_items
		WHERE payment_link_id = $1
		ORDER BY sort_order ASC, created_at ASC`, paymentLinkID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []Item
	for rows.Next() {
		var item Item
		if err := rows.Scan(&item.ID, &item.PaymentLinkID, &item.Description, &item.Quantity, &item.UnitAmount, &item.SortOrder, &item.CreatedAt); err != nil {
			return nil, err
		}
		out = append(out, item)
	}
	return out, rows.Err()
}

func (s *Service) insertItems(ctx context.Context, paymentLinkID string, items []invoicesvc.LineItemInput, currencyCode string) error {
	for i, item := range items {
		unit, err := invoicesvc.ParseFiatAmount(strings.TrimSpace(item.UnitAmount), currencyCode)
		if err != nil {
			return err
		}
		_, err = s.pool.Exec(ctx, `
			INSERT INTO payment_link_items (payment_link_id, description, quantity, unit_amount, sort_order)
			VALUES ($1, $2, $3, $4, $5)`,
			paymentLinkID, strings.TrimSpace(item.Description), strings.TrimSpace(item.Quantity), unit, i,
		)
		if err != nil {
			return err
		}
	}
	return nil
}

func (s *Service) List(ctx context.Context, organizationID, environment string, limit int) ([]PaymentLink, error) {
	if limit <= 0 {
		limit = 50
	}
	rows, err := s.pool.Query(ctx, `
		SELECT `+linkSelect+`
		FROM payment_links
		WHERE organization_id = $1 AND environment = $2
		ORDER BY created_at DESC
		LIMIT $3`, organizationID, environment, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []PaymentLink
	for rows.Next() {
		link, err := s.scanLink(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *link)
	}
	return out, rows.Err()
}

func (s *Service) ListPaginated(ctx context.Context, organizationID, environment string, query ListQuery) ([]map[string]any, int, error) {
	page := query.Page
	if page < 1 {
		page = 1
	}
	pageSize := query.PageSize
	if pageSize < 1 {
		pageSize = 20
	}
	if pageSize > 100 {
		pageSize = 100
	}
	offset := (page - 1) * pageSize
	sortOrder := strings.ToLower(query.SortOrder)
	if sortOrder != "asc" {
		sortOrder = "desc"
	}

	where := []string{"organization_id = $1", "environment = $2"}
	args := []any{organizationID, environment}
	argN := 3

	if search := strings.TrimSpace(query.Search); search != "" {
		pattern := "%" + search + "%"
		where = append(where, fmt.Sprintf(
			`(public_id ILIKE $%d OR product_name ILIKE $%d OR description ILIKE $%d)`,
			argN, argN, argN,
		))
		args = append(args, pattern)
		argN++
	}
	switch strings.TrimSpace(query.Status) {
	case "active":
		where = append(where, "active = 1")
	case "inactive":
		where = append(where, "active = 0")
	}

	whereSQL := strings.Join(where, " AND ")
	orderSQL := "created_at DESC"
	if sortOrder == "asc" {
		orderSQL = "created_at ASC"
	}

	var total int
	if err := s.pool.QueryRow(ctx, `SELECT COUNT(*) FROM payment_links WHERE `+whereSQL, args...).Scan(&total); err != nil {
		return nil, 0, err
	}

	listArgs := append(append([]any{}, args...), pageSize, offset)
	rows, err := s.pool.Query(ctx, fmt.Sprintf(`
		SELECT `+linkSelect+`
		FROM payment_links
		WHERE %s
		ORDER BY %s
		LIMIT $%d OFFSET $%d`, whereSQL, orderSQL, argN, argN+1), listArgs...)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	var links []PaymentLink
	for rows.Next() {
		link, err := s.scanLink(rows)
		if err != nil {
			return nil, 0, err
		}
		links = append(links, *link)
	}
	if err := rows.Err(); err != nil {
		return nil, 0, err
	}

	linkIDs := make([]string, 0, len(links))
	for _, link := range links {
		linkIDs = append(linkIDs, link.ID)
	}
	itemCounts := map[string]int{}
	if len(linkIDs) > 0 {
		countRows, err := s.pool.Query(ctx, `
			SELECT payment_link_id, COUNT(*)
			FROM payment_link_items
			WHERE payment_link_id = ANY($1::uuid[])
			GROUP BY payment_link_id`, linkIDs)
		if err != nil {
			return nil, 0, err
		}
		defer countRows.Close()
		for countRows.Next() {
			var id string
			var count int
			if err := countRows.Scan(&id, &count); err != nil {
				return nil, 0, err
			}
			itemCounts[id] = count
		}
		if err := countRows.Err(); err != nil {
			return nil, 0, err
		}
	}

	out := make([]map[string]any, 0, len(links))
	for i := range links {
		serialized := s.Serialize(&links[i])
		serialized["item_count"] = itemCounts[links[i].ID]
		out = append(out, serialized)
	}
	return out, total, nil
}

func (s *Service) Create(ctx context.Context, input CreateInput) (*PaymentLink, error) {
	currencyCode := invoicesvc.ResolveCurrencyCode(input.CurrencyCode)
	amount, err := invoicesvc.CalculateTotal(input.Items, currencyCode)
	if err != nil {
		return nil, err
	}

	assetConfig, err := s.paymentMethods.ResolveAssetConfig(ctx, input.OrganizationID, input.SettlementAsset, input.AllowedAssets)
	if err != nil {
		return nil, err
	}

	publicID, err := createPublicID()
	if err != nil {
		return nil, err
	}

	var productName *string
	if len(input.Items) > 0 {
		name := strings.TrimSpace(input.Items[0].Description)
		if name != "" {
			productName = &name
		}
	}

	collection := NormalizeCustomerCollection(input.CustomerCollection)
	collectionJSON, err := json.Marshal(collection)
	if err != nil {
		return nil, err
	}
	allowedJSON, err := json.Marshal(assetConfig.AllowedAssets)
	if err != nil {
		return nil, err
	}
	var metaJSON []byte
	if input.Metadata != nil {
		metaJSON, err = json.Marshal(input.Metadata)
		if err != nil {
			return nil, err
		}
	}

	var description *string
	if input.Description != nil {
		t := strings.TrimSpace(*input.Description)
		if t != "" {
			description = &t
		}
	}

	link, err := s.scanLink(s.pool.QueryRow(ctx, `
		INSERT INTO payment_links (
			public_id, organization_id, environment, amount, currency_code,
			settlement_asset, settlement_asset_issuer, allowed_assets,
			product_name, product_description, description, customer_collection,
			metadata, active
		) VALUES (
			$1, $2, $3, $4, $5,
			$6, $7, $8::jsonb,
			$9, NULL, $10, $11::jsonb,
			$12, 1
		)
		RETURNING `+linkSelect,
		publicID, input.OrganizationID, input.Environment, amount, currencyCode,
		assetConfig.SettlementAsset.AssetCode, assetConfig.SettlementAsset.IssuerAddress, allowedJSON,
		productName, description, collectionJSON, metaJSON,
	))
	if err != nil {
		return nil, err
	}
	if err := s.insertItems(ctx, link.ID, input.Items, currencyCode); err != nil {
		return nil, err
	}
	return link, nil
}

func (s *Service) Serialize(link *PaymentLink) map[string]any {
	return map[string]any{
		"id":            link.PublicID,
		"object":        "payment_link",
		"amount":        link.Amount,
		"currency_code": link.CurrencyCode,
		"settlement_asset": map[string]any{
			"asset_code":     link.SettlementAsset,
			"issuer_address": link.SettlementAssetIssuer,
		},
		"allowed_assets":      link.AllowedAssets,
		"product_name":        link.ProductName,
		"product_description": link.ProductDescription,
		"description":         link.Description,
		"customer_collection": NormalizeCustomerCollection(&link.CustomerCollection),
		"active":              link.Active,
		"metadata":            link.Metadata,
		"url":                 s.URL(link.PublicID),
		"environment":         link.Environment,
		"created_at":          link.CreatedAt,
		"updated_at":          link.UpdatedAt,
	}
}

// SerializeWithItems loads line items using the provided context.
func (s *Service) SerializeWithItems(ctx context.Context, link *PaymentLink) (map[string]any, error) {
	dbItems, err := s.ListItems(ctx, link.ID)
	if err != nil {
		return nil, err
	}
	currencyCode := ""
	if link.CurrencyCode != nil {
		currencyCode = *link.CurrencyCode
	}
	serializedItems := make([]map[string]any, 0, len(dbItems))
	for _, item := range dbItems {
		lineAmount := item.UnitAmount
		if currencyCode != "" {
			if line, err := invoicesvc.LineItemAmount(invoicesvc.LineItemInput{
				Description: item.Description,
				Quantity:    item.Quantity,
				UnitAmount:  item.UnitAmount,
			}, currencyCode); err == nil {
				lineAmount = line
			}
		}
		serializedItems = append(serializedItems, map[string]any{
			"description": item.Description,
			"quantity":    item.Quantity,
			"unit_amount": item.UnitAmount,
			"line_amount": lineAmount,
		})
	}
	out := s.Serialize(link)
	out["items"] = serializedItems
	return out, nil
}
